ext/oj/cache.c

```diff
@@ -0,0 +1,178 @@
+    }
+}
+#endif
+
+static uint32_t hash_calc(const uint8_t *key, size_t len) {
+    const uint8_t *end     = key + len;
+    const uint8_t *endless = key + (len & 0xFFFFFFFC);
+    uint32_t       h       = (uint32_t)len;
+    uint32_t       k;
+
+    while (key < endless) {
+        k = (uint32_t)*key++;
+        k |= (uint32_t)*key++ << 8;
+        k |= (uint32_t)*key++ << 16;
+        k |= (uint32_t)*key++ << 24;
+
+        k *= M;
+        k ^= k >> 24;
+        h *= M;
+        h ^= k * M;
+    }
+    if (1 < end - key) {
+        uint16_t k16 = (uint16_t)*key++;
+
+        k16 |= (uint16_t)*key++ << 8;
+        h ^= k16 << 8;
+    }
+    if (key < end) {
+        h ^= *key;
+    }
+    h *= M;
+    h ^= h >> 13;
+    h *= M;
+    h ^= h >> 15;
+
+    return h;
+}
+
+Cache cache_create(size_t size, VALUE (*form)(const char *str, size_t len), bool reg) {
+    Cache c     = ALLOC(struct _cache);
+    int   shift = 0;
+
+    for (; REHASH_LIMIT < size; size /= 2, shift++) {
+    }
+    if (shift < MIN_SHIFT) {
+        shift = MIN_SHIFT;
+    }
+    c->size  = 1 << shift;
+    c->mask  = c->size - 1;
+    c->slots = ALLOC_N(Slot, c->size);
+    memset(c->slots, 0, sizeof(Slot) * c->size);
+    c->form = form;
+    c->cnt  = 0;
+    c->reg  = reg;
+
+    return c;
+}
+
+static void rehash(Cache c) {
+    uint32_t osize = c->size;
+
+    c->size = osize * 4;
+    c->mask = c->size - 1;
+    REALLOC_N(c->slots, Slot, c->size);
+    memset(c->slots + osize, 0, sizeof(Slot) * osize * 3);
+
+    Slot *end = c->slots + osize;
+    for (Slot *sp = c->slots; sp < end; sp++) {
+        Slot s    = *sp;
+        Slot next = NULL;
+        *sp       = NULL;
+        for (; NULL != s; s = next) {
+            next = s->next;
+
+            uint32_t h      = s->hash & c->mask;
+            Slot *   bucket = c->slots + h;
+
+            s->next = *bucket;
+            *bucket = s;
+        }
+    }
+}
+
+void cache_free(Cache c) {
+    for (uint32_t i = 0; i < c->size; i++) {
+        Slot next;
+        for (Slot s = c->slots[i]; NULL != s; s = next) {
+            next = s->next;
+            xfree(s);
+        }
+    }
+    xfree(c->slots);
+    xfree(c);
+}
+
+VALUE
+cache_intern(Cache c, const char *key, size_t len) {
+    if (CACHE_MAX_KEY < len) {
+        return c->form(key, len);
+    }
+    uint32_t h      = hash_calc((const uint8_t *)key, len);
+    Slot *   bucket = c->slots + (h & c->mask);
+    Slot     b;
+    Slot     tail = NULL;
+
+    for (b = *bucket; NULL != b; b = b->next) {
+        if ((uint8_t)len == b->klen && 0 == strncmp(b->key, key, len)) {
+            return b->val;
+        }
+        tail = b;
+    }
+    b       = ALLOC(struct _slot);
+    b->hash = h;
+    b->next = NULL;
+    if (NULL == tail) {
+        *bucket = b;
+    } else {
+        tail->next = b;
+    }
+    memcpy(b->key, key, len);
+    b->klen = (uint8_t)len;
+    b->val  = c->form(key, len);
+    if (c->reg) {
+        rb_gc_register_address(&b->val);
```

Comment:
If the object passed to`rb_gc_register_address()` is no longer needed, you have to call `rb_gc_unregister_address()` to release wasted object in `cache_free()`.

Reply:
We might need to check other places that use `rb_gc_register_address()` as well to avoid memory leak.

Reply:
Good point. The cache_free will have to do that.

Reply:
Changed to rb_gc_mark. That seems to be faster as well in the grans scheme of things.

It was a bit surprising to note that the GC runs concurrently with the main thread. I wonder if a lock is needed for the parser to avoid a race condition with marking and parsing.

Reply:
Added `Oj::Parser.usual` which can be used like `Oj::Parser.usual.parse('[true]')` similar to `Oj::load`.